src/functions_framework/_typed_event.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import inspect
+
+from inspect import signature
+
+from functions_framework import _function_registry
+from functions_framework.exceptions import FunctionsFrameworkException
+
+"""Registers user function in the REGISTRY_MAP and the INPUT_TYPE_MAP.
+Also performs some validity checks for the input type of the function
+
+Args:
+    decorator_type: The type provided by the @typed(input_type) decorator
+    func: User function
+"""
+
+
+def register_typed_event(decorator_type, func):
+    try:
+        sig = signature(func)
+        annotation_type = list(sig.parameters.values())[0].annotation
+        input_type = _select_input_type(decorator_type, annotation_type)
+        _type_validity_check(input_type)
+    except IndexError:
+        raise FunctionsFrameworkException(
+            "Function signature is missing an input parameter."
+            "The function should be defined as def your_fn(in: input_type)"
+        )
+    except Exception as e:
+        raise FunctionsFrameworkException(
+            "Functions using the @typed decorator must provide "
+            "the type of the input parameter by specifying @typed(input_type) and/or using python "
+            "type annotations def your_fn(in: input_type)."
+        )
+
+    _function_registry.INPUT_TYPE_MAP[func.__name__] = input_type
+    _function_registry.REGISTRY_MAP[
+        func.__name__
+    ] = _function_registry.TYPED_SIGNATURE_TYPE
+
+
+""" Checks whether the response type of the typed function has a to_dict method"""
+def validate_return_type(response):
```

Comment:
[nit] this and the input function should probably have parallel names because they're doing similar things now. Maybe _validate_return_type and _validate_input_type?